// read-blob/dist/read-blob.js
(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.readBlob = f()}})(function(){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
/* global FileReader, Blob */

var methods = {
  arraybuffer: 'readAsArrayBuffer',
  dataurl: 'readAsDataURL',
  text: 'readAsText'
};

function readBlob (blob, type, cb) {
  var promise;

  if (typeof type !== 'string') {
    cb = type;
    type = 'arraybuffer';
  }

  if (!(blob instanceof Blob)) {
    var err = new Error('first argument is not a blob');
    if (cb) {
      cb(err);
      return;
    } else {
      return Promise.reject(err);
    }
  }

  if (!cb) {
    promise = new Promise(function (resolve, reject) {
      cb = function (err, result) {
        if (err) reject(err);
        else resolve(result);
      };
    });
  }

  var method = methods[type.toLowerCase()];
  var reader = new FileReader();

  reader.addEventListener('loadend', function onend (e) {
    reader.removeEventListener('loadend', onend);
    if (e.error) {
      cb(e.error);
    } else {
      cb(null, reader.result);
    }
  });

  if (method) {
    reader[method](blob);
  } else {
    // assume the output value is a text encoding
    reader.readAsText(blob, type);
  }

  return promise;
}

module.exports = readBlob;

// add arraybuffer, dataurl, and text methods
// as aliases to their options counterpart
Object.keys(methods).forEach(function (type) {
  readBlob[type] = function (blob, cb) {
    return readBlob(blob, type, cb);
  };
});

},{}]},{},[1])(1)
});
